cpp/ql/src/semmle/code/cpp/exprs/Expr.qll

```diff
@@ -85,10 +85,26 @@ class Expr extends StmtParent, @expr {
   override string toString() { none() }
 
   /** Gets the value of this expression, if it is a constant. */
-  string getValue() { exists(@value v | values(v,result,_) and valuebind(v,underlyingElement(this))) }
+  string getValue() { exists(@value v | values(v,result) and valuebind(v,underlyingElement(this))) }
+
+  private string getValueTextFollowingConversions(Expr e) {
+    e.getValue() = this.getValue()
+    and
+    (if exists(@value v |
+               valuebind(v,underlyingElement(e)) and
+               valuetext(v, _))
+     then exists(@value v |
+                 valuebind(v,underlyingElement(e)) and
+                 valuetext(v, result))
+     else result = this.getValueTextFollowingConversions(e.getConversion()))
+  }
```

Comment:
Do I understand it correctly that the base case of this recursion will relate all expression pairs `this` and `e` that have the same result for `getValue` and where `e` has a value text? So in a snapshot with 10,000 expressions of constant value `0`, that will be 10,000 * 10,000 tuples? That's too much.

Reply:
Good point; I restricted it to expressions that are conversions of `this`.

Reply:
The new restriction should fix the size problem, but I still don't think this predicate does what the comment says. For example, if `this` and one of its conversions `e` have different `valuetext`s, then the base case of this predicate will return both of those texts. That may not be possible with the extractor we have today, but the QL engine will still do all the extra work that's needed to account for the possibility.

I think the QLDoc you put on `getValueTextFollowingConversions` is closer to the description I put on `getExprLocationOverride` in #1781. That predicate is written to be very fast, doing work only in the case where a location needs to change and leaving the "no change" case to the non-recursive predicate that calls it. Maybe you can borrow that structure.

Can it ever happen that we need to go in the other direction because `myExpr` has `valuetext` but `(myExpr)` doesn't?